### PR TITLE
Fec 8696 contrib presets

### PR DESCRIPTION
**Pull request by esakal**

### CODE REVIEW

File: flow-typed/types/ui-options.js [hunk 1]
```diff
@@ -1,10 +1,12 @@
 // @flow
+
 declare type UIOptionsObject = {
   targetId: string,
   debugActions?: boolean,
   forceTouchUI?: boolean,
   log?: UILogConfigObject,
   components?: ComponentsConfig,
+  presetComponents?: PresetComponent[],
```
[the remark this answers is not in the document]
Comment (yairans) [reply flow-typed/types/ui-options.js]: it already got default empty array from kaltura-player https://github.com/kaltura/kaltura-player-js/pull/253/files#diff-039b70137bb24fed2504a0093f23e583R13

File: src/ui-manager.js [hunk 2]
```diff
@@ -169,13 +179,15 @@ class UIManager {
...
       const template = (
         <Provider store={this.store}>
-          <IntlProvider definition={this._translations[this._locale]}>
-            <Shell player={this.player}>
-              <EngineConnector player={this.player} />
-              <VideoPlayer player={this.player} />
-              <PlayerGUI uis={uis} player={this.player} playerContainer={this.container} />
-            </Shell>
-          </IntlProvider>
+          <ContainerProvider presetComponents={this._presetComponents}>
```
Comment (yairans) [inline src/ui-manager.js]:
```suggestion
          <ContainerProvider presetComponents={config.presetComponents}>
```
You can remove `this._presetComponents`, 
`config.presetComponents` typed as array as I mention above

File: src/components/container/container.js [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+    }
+
+    const nextContainerComponents = {
+      appendedComponents: [],
+      positionedComponentMap: {}
+    };
+    const positionedComponentMap = nextContainerComponents.positionedComponentMap;
+    let hasPositionedComponents = false;
+
+    const presetComponents = [...presetsData.allPresets, ...(presetsData.specificPreset[targetPresetName] || [])];
```
Comment (yairans) [inline src/components/container/container.js]:
```suggestion
    const presetComponents = presetsData.allPresets.concat(presetsData.specificPreset[targetPresetName] || []);
```

File: src/utils/logger.js [hunk 4]
```diff
@@ -13,6 +13,7 @@ const LogLevel: LogLevelType = {
   OFF: JsLogger.OFF
 };
 
+// TODO sakal check why not using playkit-js logger
```
Comment (OrenMe) [inline src/utils/logger.js]: In general the UI app was supposed to be playkit-js dependency free and only operate on the interface.
We tried but obviously didn't do the best-est of job, but I would like to keep it for sake of when we can define clear interface.
The real challenge here can be solved by the new bundler that can inject the logger as app wide dependency and then you avoid the duplication.
Note: you might say that the player interface has a logger, which is true and we can base upon this - open for discussion.

File: src/ui-manager.js [hunk 5]
```diff
@@ -49,9 +50,17 @@ class UIManager {
     if (config.log && config.log.level && this.LogLevel[config.log.level]) {
       setLogLevel(this.LogLevel[config.log.level]);
     }
+
+    // TODO sakal check how they currently modify player log level (probably not from ui-manager)
+    if (config.log && player && player.setLogLevel && player.LogLevel) {
+      player.setLogLevel(player.LogLevel[config.log.level]);
```
Comment (OrenMe) [inline src/ui-manager.js]: from main API on kaltura-player

File: src/player-gui.js [hunk 6]
```diff
@@ -61,7 +67,17 @@ class PlayerGUI extends Component {
     let uiToRender;
     if (this.props.uis.length > 0) {
       uiToRender = this.getMatchedUI(props.uis, props.state);
-      return uiToRender ? uiToRender.template(props) : this.props.uis[0].template(props);
+      // TODO sakal I think the fallback here is problematic and will lead to implementation issues - consider undefined instead
+      // TODO sakal  I removed the invokation of the component and left it to preact otherwise displayname are not relevant
+      const template = uiToRender ? uiToRender.template : this.props.uis[0].template;
+      const presetComponent = h(template, props);
```
Comment (OrenMe) [inline src/player-gui.js]: isn't calling `h` redundant here? is this related to the second comment? if so can;t we tell it not to mangle it? or use something else? maybe it doesn't matter if it doesn't have performance implications, does it?

regarding the fallback, never has an issue with it so can't say why it matters in hindsight.
The main issue will be for custom implementations, but keep in mind not well defined states might get stuck if you don't offer default.

File: src/utils/logger.js (hunk 4 above)
Comment (esakal) [reply src/utils/logger.js]: ok, will use this logger in my plugins

File: src/player-gui.js (hunk 6 above)
Comment (esakal) [reply src/player-gui.js]: calling `h` is required to create element with props.

regarding the fallback : changing the fallback to pick the last item to fix a bug

File: src/components/container/container-provider.js [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+      if (!component.render || !component.container) {
+        logger.warn(
+          `preset with label '${component.label ||
+            ''}' configuration is invalid, missing required configuration (did you remember to set 'container' and 'render'?)`
+        );
+        return;
+      }
+
+      if (!component.presets) {
+        allPresets.push(component);
```
[the remark this answers is not in the document]
Comment (esakal) [reply src/components/container/container-provider.js]: allow setting to all presets

File: src/player-gui.js [hunk 8]
```diff
@@ -39,6 +44,7 @@ class PlayerGUI extends Component {
    * @memberof PlayerGUI
    */
   getMatchedUI(uis: Array<any>, state: Object): any {
+    // TODO sakal did you consider performance as the conditions are dynamic? why not support it explicitly
```
[the remark this answers is not in the document]
Comment (esakal) [reply src/player-gui.js]: will be reconsidered later

File: flow-typed/types/ui-options.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (esakal) [reply flow-typed/types/ui-options.js]: done

File: src/components/shell/shell.js [hunk 9]
```diff
@@ -213,13 +213,12 @@ class Shell extends BaseComponent {
    * @memberof Shell
    */
   _onWindowResize(): void {
-    const playerContainer = document.getElementById(this.props.targetId);
+    // todo sakal use an exposed api
+    // todo sakal Oren - moved the `document.body` updateDocumentWidth to relevant listener
+    const playerContainer = document.getElementById(this.player._localPlayer._playerId);
```
Comment (OrenMe) [inline src/components/shell/shell.js]: don't access internals, not safe

File: src/ui-manager.js [hunk 10]
```diff
@@ -62,8 +149,30 @@ class UIManager {
     this.setConfig(config);
     this._setLocaleTranslations(config);
     setEnv(this.player.env);
+
+    this._resizeWatcher = new playkitUtils.ResizeWatcher();
+    this._resizeWatcher.init(document.getElementById(config.targetId));
+    this._resizeWatcher.addEventListener(CustomEventType.RESIZE, this._onPlayerWrapperResize);
```
Comment (OrenMe) [inline src/ui-manager.js]: this is thrown from player, what is the use case to hold two instances? just listen to resize from player

**Comment by esakal** [comment]
duplicate of #399

**Pull request closed without merging**